### PR TITLE
Feat/adp 1861 add tests for single address wallet shutdown

### PR DESCRIPTION
**Pull request by AngelCastilloB**
# Context

Not cleaning up the SingleAddressWallet properly has already caused issues twice. We need a unit test that collect all wallet Subjects (including in nested objects) and assert that they are all completed after calling wallet.shutdown()

# Proposed Solution

Write an unit tests that verifies all subscriptions to observables and services are cleaned up after shutdown.

# Important Changes Introduced

Fixed an issue where the Rewards service provider was not being shutdown when the wallet shutdown method was invoked.

### CODE REVIEW

**Review by mkazlauskas — APPROVED** [review]
Nice work! :rocket:

File: packages/key-management/src/util/createAsyncKeyAgent.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import { AsyncKeyAgent, KeyAgent } from '../';
 import { BehaviorSubject } from 'rxjs';
 
-export const createAsyncKeyAgent = (keyAgent: KeyAgent): AsyncKeyAgent => {
+export const createAsyncKeyAgent = (keyAgent: KeyAgent, shutdownSpy?: () => void): AsyncKeyAgent => {
```
Comment (rhyslbw) [inline packages/key-management/src/util/createAsyncKeyAgent.ts:4]:
```suggestion
export const createAsyncKeyAgent = (keyAgent: KeyAgent, onShutdown?: () => void): AsyncKeyAgent => {
```
Here's a way to ensure we're not introducing code explicitly for tests.

**Review by rhyslbw — CHANGES_REQUESTED** [review]

File: packages/key-management/src/util/createAsyncKeyAgent.ts (hunk 1 above)
Comment (rhyslbw) [reply packages/key-management/src/util/createAsyncKeyAgent.ts]: Done in https://github.com/input-output-hk/cardano-js-sdk/pull/552/commits/91b242127a566e9f052ca1fcdb9229e0c550e861

File: packages/key-management/src/util/createAsyncKeyAgent.ts [hunk 2]
```diff
@@ -16,6 +16,7 @@ export const createAsyncKeyAgent = (keyAgent: KeyAgent): AsyncKeyAgent => {
     knownAddresses$,
     shutdown() {
       knownAddresses$.complete();
+      if (onShutdown) onShutdown();
```
Comment (mkazlauskas) [inline packages/key-management/src/util/createAsyncKeyAgent.ts:19]: Alternative syntax (no need to change)
```suggestion
      onShutdown?.();
```

**Review by mkazlauskas — DISMISSED** [review]

File: packages/key-management/src/util/createAsyncKeyAgent.ts (hunk 2 above)
Comment (rhyslbw) [reply packages/key-management/src/util/createAsyncKeyAgent.ts:19]: Ah yeah, I just blindly did a rename refactor 🙈  I'll update it and squash at the same time

**Review by mkazlauskas — APPROVED** [review]

**Review by rhyslbw — APPROVED** [review]

**Pull request merged**